### PR TITLE
Hw 3

### PR DESCRIPTION
**Pull request by denys-buhlak**
Create task - 3

### CODE REVIEW

File: hw-3/src/domino/domino-block.css [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    left: 1em;
+
+    width: 1.2em;
+    height: 1.2em;
+
+    background-color: #ffffff;
+    border-radius: 4em;
+}
+
+.domino .TC135 {
```
Comment (Korew) [inline hw-3/src/domino/domino-block.css:71]: I didn't find where class `TC135` used

File: hw-3/src/style.css [hunk 2]
```diff
@@ -0,0 +1,130 @@
+:root {
```
Comment (Korew) [inline hw-3/src/style.css:1]: you have 4 different css files and a lot of code repeats. It's ok to make one css for all three tasks. Or you could create one general css file + 3 other files with unique styles. On main page you link only general css, on other pages general + unique css.

File: hw-3/src/domino/domino.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    let index = array.indexOf(element);
+
+    while (index != -1) {
+        indexes.push(index);
+        index = array.indexOf(element, index + 1);
+    }
+    
+    return (indexes);
+}
```
Comment (Korew) [inline hw-3/src/domino/domino.js:125]: I couldn't check domino task - I clicked `start` button maybe hundred times and all the time it said `Imposible. Try again!`. Please add inputs, where I can enter my own numbers

**Review by Korew — CHANGES_REQUESTED** [review]
Deploy to netlify - very good, thank you :)

Please, prettify your code. It should be perfect - no unneeded empty lines, no double spaces (but in some cases you miss spaces), no trailing spaces etc. Read comments and apply changes all over the project.

 I was not able to check domino task - add inputs so I can enter my own numbers. 
 
Try to break task into several simple subtasks and write simple functions. Each function should do one action. For example for coins task you can create two functions - one will draw a line. Second function will draw whole tree (in second you will call first function as many times as many lines you have). 
 
Use only English in your code :)

File: hw-3/src/coins/coins.css [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    line-height: var(--line-height-main);
+    font-weight: var(--font-weight-main);
+    font-size: var(--font-size-main);
+
+    font-style: var(--font-style-main);
+    text-align: left;
+}
+
+.container {
+    width: 100%;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.css]: fix

File: hw-3/src/coins/coins.css [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+.site-footer {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    min-height: 60px;
+}
+
+.site-footer p {
+    width: 100%;
+
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.css]: fix

File: hw-3/src/atm/atm.html [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+        <meta name="viewport"
+              content="width=device-width, initial-scale=1.0">
+
+        <title>Home work JS | atm</title>
+
+        <link rel="stylesheet"
+              href="atm.css">
+
+    </head>
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.html]: fix

File: hw-3/src/atm/atm.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
+function giveMyMoney(requestSum, notes) {
+    let needNotes = [];
+    let minNotes  = [];
+
+    minNotes [0] = 0;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: question

File: hw-3/src/coins/coins.html [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        <meta name="viewport"
+              content="width=device-width, initial-scale=1.0">
+
+        <title>Home work JS | coins</title>
+
+        <link rel="stylesheet"
+              href="coins.css">
+
+    </head>
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.html]: fix

File: hw-3/src/domino/domino-block.css [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    width: 6em;
+    height: 12em;
+    margin: 10px;
+
+    font-size: 10px;
+
+    background-color: #000000;
+
+    border-radius: 0.8em;
+
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/domino/domino-block.css]: fix
each empty line separate logic block

File: hw-3/src/domino/domino-block.css (hunk 1 above)
Comment (denys-buhlak) [reply hw-3/src/domino/domino-block.css]: fix
Delete

File: hw-3/src/style.css (hunk 2 above)
Comment (denys-buhlak) [reply hw-3/src/style.css:1]: fix

File: hw-3/src/domino/domino.js (hunk 3 above)
Comment (denys-buhlak) [reply hw-3/src/domino/domino.js]: fix
Random is awesome ^)

File: hw-3/src/atm/atm.html [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+                                   id="nominal-50"
+                                   name="nominal-50"
+                                   type="number"
+                                   min="0"
+                                   max="100"
+                                   step="10"
+                                   title="enter nominals (min=0 max=100)"
+                                   data-nominal="50">
+                        </label>
+                        <label for="nominal-100">100:
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.html]: fix

File: hw-3/src/atm/atm.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+    // перебираем все суммы до заданной суммы (можно начать с мин.купюры)
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    for (let sum = 1; sum <= requestSum; sum++) {     
+
+        // по умолчанию эта сумма не для выдачи
+        minNotes [sum] = Number.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix, but not sure. I correct eslint rules

File: hw-3/src/atm/atm.js [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    }
+ 
+    // Если заданную сумму нельзя разложить по заданным купюрам
+    if (minNotes [requestSum] === Number.MAX_VALUE) { return []; }
+ 
+    // minNotes  - хранит минимальное кол-во купюр, которые нужны для выдачи заданной суммы
+    let sum = requestSum;
+
+    while (sum > 0) {
+        let curSum = sum;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+            let isNoteExist = notes[note].count > 0;
+
+            if (isNoteExist && sum >= notes[note].val && (minNotes [sum] == minNotes [sum - notes[note].val] + 1 || minNotes [sum] == minNotes [sum - notes[note].val])) {                                  
+                if (!needNotes[note]) {
+                    needNotes[note] = 0;
+                }
+                ++needNotes[note];  
+                sum -= notes[note].val;
+                notes[note].count -= 1;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+        if (curSum === sum) {
+            needNotes = [];
+            break; // не хватает купюр
+        }
+    }
+    return needNotes;
+}
+
+function getCash (notes,requestSum) {
+    const needNotes = giveMyMoney(requestSum, notes);
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+    return {outputMessage, notes};
+
+}
+
+Object.prototype.clone = function () {
+    let newObj = (this instanceof Array) ? [] : {};
+
+    for (let i in this) {
+        if (i == 'clone')
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+});
+
+/* КНОПКА СНЯТИЕ НАЛИЧНЫХ */
+document.querySelector('#getCash').addEventListener('click', (event) => {
+    event.preventDefault();
+    const requestSum = +document.querySelector('#get-cash').value;
+    const maxSum = document.querySelector('input[name="total-sum"]').value;
+    let startNotes = getValuesFromInputs(); 
+    let notes=startNotes.clone();
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: not sure

File: hw-3/src/coins/coins.js [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+        for(let j = 0; j < i  ; j++) {
+            lineForDraw += symbolForDraw;
+            quantityItemsUses=i;
+            
+        }
+        lineForDrawTotal+=lineForDraw+'\n';
+        quantityItemsRemainder=quantityItemsRemainder-quantityItemsUses;
+      
+        // console.log(lineForDraw);
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    if (quantityItemsRemainder != 0) {
+        
+        lineForDrawLast=symbolForDraw.repeat(quantityItemsRemainder);
+        console.log('\n' + lineForDrawLast);
+        lineForDrawTotal=lineForDrawTotal + 'Remainder (' + quantityItemsRemainder + '):'+ lineForDrawLast;
+     
+    }
+    console.info('init Items:'+quantintyItemsInit,'\nfull Row:' + quantityItemsUses,'\nquantity Items Remainder:' + quantityItemsRemainder);    
+    return ({
+        quantintyItemsInit:quantintyItemsInit,
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        quantityItemsUses:quantityItemsUses,
+        quantityItemsRemainder:quantityItemsRemainder,
+        lineForDrawTotal:lineForDrawTotal
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    let lineForDrawLast='';
+
+    quantityItemsRemainder = quantintyItemsInit;
+    console.clear();
+
+    // if (quantintyItemsInit < 100) {
+    //     console.log(lineForDraw);
+    // } else {
+    //     console.log('too many for draw');
+    // }
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+    let quantityItemsUses=0;
+    let quantityItemsRemainder=0;
+
+    let lineForDraw='';
+    let lineForDrawTotal='';
+    let lineForDrawLast='';
+
+    quantityItemsRemainder = quantintyItemsInit;
+    console.clear();
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+     
+    }
+    console.info('init Items:'+quantintyItemsInit,'\nfull Row:' + quantityItemsUses,'\nquantity Items Remainder:' + quantityItemsRemainder);    
+    return ({
+        quantintyItemsInit:quantintyItemsInit,
+        quantityItemsUses:quantityItemsUses,
+        quantityItemsRemainder:quantityItemsRemainder,
+        lineForDrawTotal:lineForDrawTotal
+    });
+}
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        lineForDrawTotal+=lineForDraw+'\n';
+        quantityItemsRemainder=quantityItemsRemainder-quantityItemsUses;
+      
+        // console.log(lineForDraw);
+        
+    }
+    console.log(lineForDrawTotal);
+
+    if (quantityItemsRemainder != 0) {
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+
+    outputElement.innerText=sA.lineForDrawTotal;
+    descriptionElement.innerText='Init items:' + sA.quantintyItemsInit + '\nFull rows:' + sA.quantityItemsUses + '\nItems Remainder:' + sA.quantityItemsRemainder;
+   
+});
+
+function startAction(quantintyItemsInit=0, symbolForDraw = '$') {
+
+    let quantityItemsUses=0;
+    let quantityItemsRemainder=0;
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/coins/coins.js [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    quantityItemsRemainder = quantintyItemsInit;
+    console.clear();
+
+    // if (quantintyItemsInit < 100) {
+    //     console.log(lineForDraw);
+    // } else {
+    //     console.log('too many for draw');
+    // }
+
+    for (let i = 1; i <=  quantityItemsRemainder  ; i++) {
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/coins/coins.js]: fix

File: hw-3/src/atm/atm.js [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+
+        if (curSum === sum) {
+            needNotes = [];
+            break; // не хватает купюр
+        }
+    }
+    return needNotes;
+}
+
+function getCash (notes,requestSum) {
```
[the remark this answers is not in the document]
Comment (denys-buhlak) [reply hw-3/src/atm/atm.js]: fix

File: hw-3/src/atm/atm.js [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+const getInfo = (notes, cash) => {
+  if (cash) {
+    for (const nominal of cash) {
+      notes[nominal] -= 1;
+    }
+
+    return notes;
+  }
+
+  return null;
```
Comment (Korew) [inline hw-3/src/atm/atm.js:104]:
```
  if (!cash) return null;
  
  for (const nominal of cash) {
    notes[nominal] -= 1;
  }
  return notes;
```

This is absolutely same code, but mine version 2 lines shorter :)

File: hw-3/src/coins/coins.js [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+  const sA = startAction(coinsTotal, symbolForDraw);
+
+  outputElement.innerText = sA.lineForDrawTotal;
+  descriptionElement.innerText =
+ `Init items:${sA.coinsTotal}\nFull rows:${sA.coinsUses}\nItems Remainder:${sA.coinsLeft}`;
+
+  return null;
+};
+
+startButton.addEventListener('click', (event) => {
```
Comment (Korew) [inline hw-3/src/coins/coins.js:58]: `'click', event => {`
if you have 1 argument - brackets not needed

**Review by Korew — APPROVED** [review]
Great!

**Pull request merged**